materialize-mongodb/driver.go

```diff
@@ -0,0 +1,201 @@
+package main
+
+import (
+  "context"
+  "encoding/json"
+  "fmt"
+
+  schemagen "github.com/estuary/connectors/go-schema-gen"
+  boilerplate "github.com/estuary/connectors/materialize-boilerplate"
+  pf "github.com/estuary/flow/go/protocols/flow"
+  pm "github.com/estuary/flow/go/protocols/materialize"
+  //log "github.com/sirupsen/logrus"
+
+  "go.mongodb.org/mongo-driver/mongo"
+  "go.mongodb.org/mongo-driver/mongo/options"
+  "go.mongodb.org/mongo-driver/mongo/readpref"
+)
+
+// driver implements the DriverServer interface.
+type driver struct{}
+
+func (d *driver) Connect(ctx context.Context, cfg config) (*mongo.Client, error) {
+  // Create a new client and connect to the server
+  client, err := mongo.Connect(ctx, options.Client().ApplyURI(cfg.ToURI()))
+  if err != nil {
+    return nil, err
+  }
+
+  // Ping the primary
+  if err := client.Ping(ctx, readpref.Primary()); err != nil {
+    return nil, err
+  }
+
+  return client, nil
+}
+
+func (d driver) Spec(ctx context.Context, req *pm.SpecRequest) (*pm.SpecResponse, error) {
+  if err := req.Validate(); err != nil {
+    return nil, fmt.Errorf("validating request: %w", err)
+  }
+
+  es := schemagen.GenerateSchema("Materialize MongoDB Spec", &config{})
+  endpointSchema, err := es.MarshalJSON()
+  if err != nil {
+    return nil, fmt.Errorf("generating endpoint schema: %w", err)
+  }
+
+  resourceSchema, err := schemagen.GenerateSchema("MongoDB Collection", &resource{}).MarshalJSON()
+  if err != nil {
+    return nil, fmt.Errorf("generating resource schema: %w", err)
+  }
+
+  return &pm.SpecResponse{
+    EndpointSpecSchemaJson: json.RawMessage(endpointSchema),
+    ResourceSpecSchemaJson: json.RawMessage(resourceSchema),
+    DocumentationUrl:       "https://go.estuary.dev/materialize-mongodb",
+    Oauth2Spec:             nil,
+  }, nil
+}
+
+func (d driver) Validate(ctx context.Context, req *pm.ValidateRequest) (*pm.ValidateResponse, error) {
+  cfg, err := resolveEndpointConfig(req.EndpointSpecJson)
+  if err != nil {
+    return nil, err
+  }
+
+  _, err = d.Connect(ctx, cfg)
+  if err != nil {
+    return nil, fmt.Errorf("connecting to database: %w", err)
+  }
+
+  // MongoDB supports arbitrary JSON documents so there are no constraints on
+  // any fields and all fields are marked as "recommended"
+  var out []*pm.ValidateResponse_Binding
+  for _, b := range req.Bindings {
+    res, err := resolveResourceConfig(b.ResourceSpecJson)
+    if err != nil {
+      return nil, err
+    }
+
+    constraints := make(map[string]*pm.Constraint)
```

Comment:
If we're only materializing the flow doc, I think we should require the flow document be included in the projections and actually forbid anything else. This is how the materializations for [pubsub](https://github.com/estuary/connectors/blob/2a7aa9c6a5d7432998859cd8706c1d36240af453/materialize-google-pubsub/driver.go#L143-L155) and [not-really-a-materialization webhooks](https://github.com/estuary/connectors/blob/2a7aa9c6a5d7432998859cd8706c1d36240af453/materialize-webhook/driver.go#L95-L107) do it. This would be more explicit than saying that any field is recommended but then ignoring that set of selected fields and always materializing the entire document anyway.

Reply:
Addressed this as well 👍🏽